l10n_it_fiscalcode/model/res_partner.py

```diff
@@ -1,31 +1,93 @@
+    is_soletrader = fields.Boolean(
+        string='Sole Trader',
+        default=False,
+        readonly=False,
+        help="Checked if company is a sole trader")
+
     @api.multi
-    def check_fiscalcode(self):
+    @api.onchange('is_company')
+    def onchange_iscompany(self):
+        """ if partner is switched from company to person,
+        is_soletrader is set to False because a simple private citizen
+        may not be a sole trader.
+        A proper warning is displayed.
+        """
         for partner in self:
-            if not partner.fiscalcode:
-                return True
-            elif len(partner.fiscalcode) != 16 and partner.individual:
-                return False
-            else:
-                return True
+            if partner.is_company is False:
+                partner.is_soletrader = False
+                title = _('Partner type changed')
+                message = _('WARNING:\n'
+                            'Changing partner type from company to '
+                            'person will remove the "Sole Trader" selection '
+                            'in the Accounting tab.\nFiscal code may need '
+                            'to be changed accordingly.')
+                result = {
+                    'warning': {'title': title,
+                                'message': message, }, }
+                return result
 
-    fiscalcode = fields.Char(
-        'Fiscal Code', size=16, help="Italian Fiscal Code")
-    individual = fields.Boolean(
-        'Individual', default=False,
-        help="If checked the C.F. is referred to a Individual Person")
-
-    _constraints = [
-        (check_fiscalcode,
-         "The fiscal code doesn't seem to be correct.", ["fiscalcode"])
-    ]
+    # Constraints and onchanges
+    @api.multi
+    @api.constrains('fiscalcode', 'is_soletrader')
+    def _check_fiscalcode_constraint(self):
+        """ verify fiscalcode content, lenght and isnumeric/isalphanum
+        depending if partner is private citizen,
+        business company or sole trader
+        """
+        for partner in self:
+            if not partner.fiscalcode:
+                # fiscalcode empty. Nothing to check..
+                is_fc_ok = True
+            elif (not partner.is_company or
+                  partner.is_company and partner.is_soletrader):
+                # partner is a private citizen
+                # or a sole trader operating in Italy
+                # should have an Italian valid fiscalcode
+                if is_valid(partner.fiscalcode):
+                    is_fc_ok = True
+                else:
+                    is_fc_ok = False
+                    msg = _("The Fiscal Code is invalid")
+            elif partner.is_company and not partner.is_soletrader:
+                # partner is a business company
+                # should have the fiscal code of the same kind of VAT code
+                if (partner.fiscalcode.isnumeric() and
```

Comment:
better use of isdigit instead of isnumeric ....
read this:
http://stackoverflow.com/questions/24384852/difference-between-unicode-isdigit-and-unicode-isnumeric